ddtrace/tracer/writer.go

```diff
@@ -0,0 +1,305 @@
+	// the JSON.
+	logBufferSuffix = "]}\n"
+
+	// logBufferLimit is the maximum size log line allowed by cloudwatch
+	logBufferLimit = 256 * 1024
+)
+
+func (h *logTraceWriter) resetBuffer() {
+	h.buf.Reset()
+	h.buf.WriteString(`{"traces": [`)
+	h.hasTraces = false
+}
+
+// encodeFloat correctly encodes float64 into the JSON format followed by ES6.
+// This code is reworked from Go's encoding/json package
+// (https://github.com/golang/go/blob/go1.15/src/encoding/json/encode.go#L573)
+//
+// One important departure from encoding/json is that infinities and nans are encoded
+// as null rather than signalling an error.
+func encodeFloat(p []byte, f float64) []byte {
+	if math.IsInf(f, 0) || math.IsNaN(f) {
+		return append(p, "null"...)
+	}
+	abs := math.Abs(f)
+	if abs != 0 && (abs < 1e-6 || abs >= 1e21) {
+		p = strconv.AppendFloat(p, f, 'e', -1, 64)
+		// clean up e-09 to e-9
+		n := len(p)
+		if n >= 4 && p[n-4] == 'e' && p[n-3] == '-' && p[n-2] == '0' {
+			p[n-2] = p[n-1]
+			p = p[:n-1]
+		}
+	} else {
+		p = strconv.AppendFloat(p, f, 'f', -1, 64)
+	}
+	return p
+}
+
+func (h *logTraceWriter) encodeSpan(s *span) {
+	var scratch [maxFloatLength]byte
+	h.buf.WriteString(`{"trace_id":"`)
+	h.buf.Write(strconv.AppendUint(scratch[:0], uint64(s.TraceID), 16))
+	h.buf.WriteString(`","span_id":"`)
+	h.buf.Write(strconv.AppendUint(scratch[:0], uint64(s.SpanID), 16))
+	h.buf.WriteString(`","parent_id":"`)
+	h.buf.Write(strconv.AppendUint(scratch[:0], uint64(s.ParentID), 16))
+	h.buf.WriteString(`","name":"`)
+	h.buf.WriteString(s.Name)
+	h.buf.WriteString(`","resource":"`)
+	h.buf.WriteString(s.Resource)
+	h.buf.WriteString(`","error":`)
+	h.buf.Write(strconv.AppendInt(scratch[:0], int64(s.Error), 10))
+	h.buf.WriteString(`,"meta":{`)
+	first := true
+	for k, v := range s.Meta {
+		if first {
+			h.buf.WriteByte('"')
+			first = false
+		} else {
+			h.buf.WriteString(`,"`)
+		}
+		h.buf.WriteString(k)
+		h.buf.WriteString(`":"`)
+		h.buf.WriteString(v)
+		h.buf.WriteString(`"`)
+	}
+	h.buf.WriteString(`},"metrics":{`)
+	first = true
+	for k, v := range s.Metrics {
+		if math.IsNaN(v) || math.IsInf(v, 0) {
+			// The trace forwarder does not support infinity or nan, so we do not send metrics with those values.
+			continue
+		}
+		if first {
+			h.buf.WriteByte('"')
+			first = false
+		} else {
+			h.buf.WriteString(`,"`)
+		}
+		h.buf.WriteString(k)
+		h.buf.WriteString(`":`)
+		h.buf.Write(encodeFloat(scratch[:0], v))
+	}
+	h.buf.WriteString(`},"start":`)
+	h.buf.Write(strconv.AppendInt(scratch[:0], s.Start, 10))
+	h.buf.WriteString(`,"duration":`)
+	h.buf.Write(strconv.AppendInt(scratch[:0], s.Duration, 10))
+	h.buf.WriteString(`,"service":"`)
+	h.buf.WriteString(s.Service)
+	h.buf.WriteString(`"}`)
+}
+
+type encodingError struct {
+	cause      error
+	dropReason string
+}
+
+// writeTrace makes an effort to write the trace into the current buffer. It returns
```

Comment:
The documentation and inline comments for this function are great 👍 Thanks for doing that!